[WIP] fs: feature detection for recursive mkdir[Sync]

MylesBorins opened the pull request:
~~This commit adds a non-enumerable and non-writable property
`Symbol('recursive')` to fs.mkdir and fs.mkdirSync to allow developers
to feature detect~~

This PR adds a known symbol `util.features` that can be used to do feature detection on particular APIS.

How it can be used:

edited:

```js
fs.mkdir[util.features].recursive
// true
```

addaleax commented:
I don’t think we should encourage that kind of Symbol-digging, and I don’t really see a reason for this to be non-enumerable (or non-writable, since writability might be nice for tests).

What’s the motivation for making it so hard, instead of having `fs.mkdir.hasRecursive = true` and `fs.mkdirSync.hasRecursive = true`?

cjihrig commented on doc/api/fs.md:
````diff
@@ -2086,6 +2086,16 @@ fs.mkdir('/tmp/a/apple', { recursive: true }, (err) => {
+can be used to detect if the recursive feature is available
+
+```js
+Object.getOwnPropertySymbols(fs.mkdir).map((sym) => {
````
What if instead of a symbol specific to this feature we exposed a "feature" symbol? It could be reused in other methods, and would be well known, so it wouldn't require looping over properties.

MylesBorins replied:
Sounds interesting! Are you imagining that `Symbol('features')` would then have an array with a list of all features? Can you share a code example of how userland feature detection would look?

cjihrig replied:
I was thinking that `Symbol('features')` would be public API, which is how users would know to access it. Then, I'd do something like:

```js
fs.mkdir[kFeaturesSymbol] = { recursive: true };
```

Open to suggestions on that implementation, but I'd go with an object over an array because it's a bit more flexible to work with IMO.

From a user's perspective:

```js
const features = `require('util').FEATURES_SYMBOL;` // also open to suggestions here.
const { mkdir } = require('fs');

if (typeof features == 'symbol' && mkdir[features].recursive) {
  // use the new thing
} else {
  // do the old thing
}
```

MylesBorins replied:
Just pushed a commit to take this approach, lmk what you think

MylesBorins commented:
@addaleax I was thinking that this would be a good place to start for a discussion about how to do feature detection. The biggest reason for making this a symbol and non-enumerable rather than just an extension to the object was that this method is unlikely to break any current workflows or monkey patching of the object. This could scale across the entire code base without disrupting or extending any object in a noticeable way.

While the symbol digging is a bit verbose it could easily be abstracted into a user land lib to

```js
function hasFeature(obj, feature) {
  Object.getOwnPropertySymbols(obj).map((sym) => {
    return sym.toString();
  }).includes(`Symbol(${feature})`);
}

```

devsnek commented:
we really should have just made a separate method

MylesBorins commented:
@devsnek it hasn't landed in a release yet. If you think that is the case we still have time to rebase it out of 10.x and land another change. That being said I disagree with the sentiment and don't feel that we need another API just for this case.

Feature detection is a larger problem in core, and I personally like the idea of using symbols for features as opposed to sniffing semver or for specific bits on the object. If we follow @cjihrig's suggestion about and make a `features` symbol, we could potentially share that symbol as a global which will make getting the results far less complicated

```js
// in common
common.features = Symbol('features');

// in fs

Object.defineProperties(mkdir, {
  common.features: ['recursive']
});

// in userland

fs.mkdir[common.features].includes('recursive');
```

boneskull commented:
What about `process.features` or `process.config`?  Would `process.features.fs_mkdirp = true` make sense?

boneskull commented:
Check for *capabilities*, not for versions.  :-1: to any solution involving SemVer

MylesBorins commented:
Quick update... I've taken @cjihrig's suggestion and made a well known symbol exposed on util and made the value an object rather than an array... for the most part this hides all the symbol bits and makes the userland code much more readable

```js
fs.mkdir[util.features].recursive;
// true
```

@addaleax does this alleviate some of your concerns?

jdalton commented on doc/api/util.md:
```diff
@@ -179,6 +179,13 @@ The `--throw-deprecation` command line flag and `process.throwDeprecation`
+-->
+
+* {symbol} that can be used to do feature detection.
+
```
😍😻😍

addaleax commented on doc/api/fs.md:
````diff
@@ -2086,6 +2086,14 @@ fs.mkdir('/tmp/a/apple', { recursive: true }, (err) => {
 ```
 
+The `util.features` symbol can be used to feature detect if
+recusion is available.
````
typo: recursion

MylesBorins replied:
I wanted to recuse myself 😅

addaleax commented on doc/api/fs.md:
```diff
@@ -2106,6 +2114,14 @@ changes:
 This is the synchronous version of [`fs.mkdir()`][].
 
+The `util.features` symbol can be used to feature detect if
+recusion is available.
```
ditto

MylesBorins commented:
I've added this to the `tsc-agenda` as this would set a precedent for how we would do feature detection in core.

richardlau commented:
As @boneskull [pointed out](https://github.com/nodejs/node/pull/22302#issuecomment-412682440) we already have `process.features` (since at least v0.10.x but probably earlier). I don't know if anyone is actually using it for feature detection (we appear to have not documented it) but that is what it appears to be for. (I can't recall without checking what's currently in there -- I think there were some openssl features (since removed?) at one point).

Is this PR a better alternative? Possibly. I think `process.features` makes sense for larger coarser grained features not necessarily for method level tweaks (I can imagine `process.features` becoming overly large if we went down this route).

cjihrig commented:
> I think process.features makes sense for larger coarser grained features not necessarily for method level tweaks (I can imagine process.features becoming overly large if we went down this route).

I agree with that statement.

cjihrig commented on doc/api/util.md:
```diff
@@ -179,6 +179,13 @@ The `--throw-deprecation` command line flag and `process.throwDeprecation`
 
+### util.features
+<!-- YAML
+added: REPlACEME
```
l -> L

devsnek commented on lib/util.js:
```diff
@@ -1472,6 +1472,7 @@ module.exports = exports = {
   callbackify,
   debuglog,
   deprecate,
+  features: Symbol('features'),
```
can the description be `node.util.features`

MylesBorins replied:
I'm not sure what you mean

RyanZim commented:
One drawback of this approach is that users will need to do recursive nesting checks to avoid getting errors in older Node versions.

For this feature specifically, you'd have to do:
```js
util.features && fs.mkdir[util.features].recursive
```
However, in the future, it gets even worse; if you're checking for hypothetical `modulex.methody[util.features].featurez` that got added in some future release, and you have to support Node versions back before the addition of `util.features`, you have to do three checks:
```js
util.features && modulex.methody[util.features] && modulex.methody[util.features].featurez
```
It works, but it's a pretty long-winded feature detection check IMO.

Fishrock123 commented on doc/api/fs.md:
```diff
@@ -2106,6 +2114,14 @@ changes:
 Synchronously creates a directory. Returns `undefined`.
 This is the synchronous version of [`fs.mkdir()`][].
 
+The `util.features` symbol can be used to feature detect if
```
Link to `util.features`?

Fishrock123 commented:
> ```
> util.features && modulex.methody[util.features] && modulex.methody[util.features].featurez
> ```

I don't see why the first conditional is necessary.

This should work fine + cache:

```
const features = modulex.methody[util.features]
if (features && features.recursive) { }
```

MylesBorins commented:
@RyanZim

Alternative approach and less verbose

```
try {
  if (!fs.mkdir[util.features].recursive) throw 'woops'
  // Do a thing
}
catch (e) {
  // Polyfill
}
```

jasnell commented:
TBH, I'm not a fan of this approach when we could just have a separate `if (fs.mkdirp)` to do the same thing.

MylesBorins commented:
@jasnell it doesn't seem scalable to me to make a new API for every new flag or feature we add to an API

ChALkeR commented:
I don't like this approach. While feature detection is a good thing to have, and I am not against landing this per se, this **does not** solve the recursive mkdir problem. And I am against landing this as a solution to the recursive mkdir feature detection problem.

The thing is -- users (ok, _almost all_ users) won't care about complex feature detection. If they care about old versions -- they will use userland `mkdirp` module. If they don't care about old Node.js versions -- they will just use the native impl. So they will just use the most simple way to poke the new code path, which currently is `fs.mkdir(path, {recursive: true})`, without even caring about feature decetion stuff.

And this is what would happen old old versions (e.g. when someone would try to use old Node.js to run a module which expects new API):
```js
> fs.mkdir('./foo/sdfsd/df', { recursive: true }, (...args) => console.log(args))
undefined
> [ { Error: ENOENT: no such file or directory, mkdir './foo/sdfsd/df'
    errno: -2,
    code: 'ENOENT',
    syscall: 'mkdir',
    path: './foo/sdfsd/df' } ]
```
This error message is completely unhelpful for the end user, and looks like there is a bug somewhere, either in their code or their deps.

With `fs.mkdirp`, this is what would happen:

```js
> fs.mkdirp('./foo/sdfsd/df', (...args) => console.log(args))
TypeError: fs.mkdirp is not a function
```

Why not just a one-line `fs.mkdirp` wrap alias to `fs.mkdir`, as @jasnell suggests above? We already have aliases of different sorts. See previous discussion at https://github.com/nodejs/node/pull/21875#issuecomment-410482206 (and response to that).

It would be shorter, easier to support, would involve less code and doc changes, and will provide both the ability to feature-detect and a more simple way to use it while giving sensible errors to users of unsupported Node.js versions.

apapirovski commented:
I'm not even convinced it should be an alias (but rather mkdirp should be the primary interface). But in general agreement with @jasnell and @ChALkeR here.

cjihrig commented:
> this does not solve the recursive mkdir problem

I don't think that is a valid statement - more of an opinion. It's just a matter of whether or not users would choose to use the feature detection or not.

To be honest, I think the function should have landed as `mkdirp()`, and it's not too late for that. However, this PR is larger than `mkdirp()` IMO. We will keep running into this same issue any time an API change is made that cannot be programmatically detected.

ChALkeR commented:
@cjihrig 
>> this does not solve the recursive mkdir problem
> I don't think that is a valid statement - more of an opinion. It's just a matter of whether or not users would choose to use the feature detection or not.

I am not sure what you mean by this. I think that the difference is objective in a sense would be observable in the ecosystem.

What problem would this solve? I.e. would they use feature detection **for**?

They won't use it to decide if they need `mkdirp` module -- if they already have it in deps, they will just use it, that's more simple.

They will not use to choose between native and their own impl -- if they expect the code to be run on older Node.js versions, they would just use `mkdirp`.

They would likely not use it to print a sensible error message -- too much code for that. Either `mkdirp` or just specifying `engines` and closing invalid issues.

The only realistic usecase for this feature detection is _inside_ the `mkdirp` module. And, imo, just that isn't even worth it.

On the other hand, `fs.mkdirp` would be less code and docs, and would provide end users with a more sensible error message when something would go wrong. It will partially solve a real problem out there that would affect users trying to run code relying on native recursive mkdir on older Node.js versions --  it will make their error messages saner and would mean less invalid bugreports in libraries out there.

And it would also solve the feature detection, for whomever needs it (which is `mkdirp`).

RyanZim commented:
@Fishrock123 you're right, if `util.features` is `undefined`, we can still do `modulex.methody[util.features]` safely. Sorry, just wasn't thinking this through.

---

> The thing is -- users (ok, almost all users) won't care about complex feature detection. If they care about old versions -- they will use userland mkdirp module.

Good point, I'm thinking about feature detection since I'm a maintainer of `fs-extra`, and I'd like to use the native option where possible (falling back to userland `mkdirp`) without having to pull in `semver` to do detection.

I still prefer a `recursive` option over a new method, it feels weird to have a separate method that just does slightly different behavior. Also, adding a `mkdirp` method will introduce a name conflict with current versions of `fs-extra`, which will result in native `mkdirp` being overwritten by our implementation. I realize this is a userland problem, not a core problem, but it's worth bringing up since there's 29M downloads of `fs-extra` per month.

---

Personally, I'd prefer the simple solution of `fs.mkdir.supportsRecursive` set to `true`. This allows a simple check, it will be `undefined` in older versions (which is falsy), and doesn't require any additions to `util` (or an extra `require('util')` just to perform the check).

cjihrig commented:
> What problem would this solve? I.e. would they use feature detection for?

This solves the exact same problem as checking for the existence of `fs.mkdirp()` in the alternative proposal 😄 

> They won't use it to decide if they need mkdirp module -- if they already have it in deps, they will just use it, that's more simple.

By that logic, they wouldn't bother using `fs.mkdirp()` either.

> They would likely not use it to print a sensible error message

While slightly more verbose, the check that @Fishrock123 wrote in https://github.com/nodejs/node/pull/22302#issuecomment-412906762 would be equivalent to checking for the existence of `fs.mkdirp()`.

> The only realistic usecase for this feature detection is inside the mkdirp module.

The use case is identical to checking for the existence of `fs.mkdirp()`, but can be extended to other new APIs as they are added. In other words, it's more scalable - even though in this particular case I'd prefer `fs.mkdirp()`.

vsemozhetbyt commented on doc/api/util.md:
```diff
@@ -179,6 +179,13 @@ The `--throw-deprecation` command line flag and `process.throwDeprecation`
 property take precedence over `--trace-deprecation` and
 `process.traceDeprecation`.
 
+### util.features
```
Nit: `###` -> `##`

boneskull commented:
> They won't use it to decide if they need mkdirp module -- if they already have it in deps, they will just use it, that's more simple.

They will, yes--I think this was expected.  Eventually, though, they won't.

boneskull commented:
@RyanZim `fs-extra` monkeypatches the `fs` module?  I didn't realize that.

I don't *think* a conflict with `fs.mkdirp` is a problem, since anyone using `fs-extra` should (*should*) know that this is happening, and won't expect to use core's implementation/API.  I'd be more concerned if it was the other way around (core's implementation blasted `fs-extra`'s).

@MylesBorins 

> it doesn't seem scalable to me to make a new API for every new flag or feature we add to an API

By this do you mean adding more properties to `process.features`?  I wouldn't say every new feature needs a property; just the ones where it'd be useful, and we can take that on a case-by-case basis.  

`mkdirp` is one of those, and a forthcoming `rimraf` implementation will have a similar issue if it reuses `fs.rmdir`--though in this case, the arity of that function is likely to change.

bcoe commented:
Having slept on it, I'm a fan of @MylesBorins' approach:

> I don’t really see a reason for this to be non-enumerable:

I'm in agreement with @MylesBorins, I think it's better to be defensive in the implementation and keep `fs.mkdir` as close as possible to the old signature as possible.

I also think we should avoid making the property writeable. The danger being that a dependency of a dependency mucks with the property, causing bizarre behavior up the stack. If someone needs to mock the behavior they can potentially mock the `require('fs').mkdir` instead?

> TBH, I'm not a fan of this approach when we could just have a separate if (fs.mkdirp)

I agree with @MylesBorins that this will potentially lead to the file system API growing too much. While it might be nice for `mkdirp`, I'm specifically thinking of @boneskull's work on `rm -r -f`, where I imagine both `force` and `recursive` would be additional flags.

bcoe commented on test/parallel/test-fs-mkdir.js:
```diff
@@ -172,6 +173,14 @@ if (common.isMainThread && (common.isLinux || common.isOSX)) {
 
+// mkdirp and mkdirSyncp feature detection
+{
+  assert.strictEqual(fs.mkdir[features].recursive, true);
```
let's add a test for the `undefined` case?

MylesBorins replied:
which undefined case? where `features` is undefined?

bcoe replied:
yeah, I think it would be good to demonstrate that this will work fine for feature detection when `util.features` is `undefined`.

```
!!mkdir[require('util').propretyThatDoesntExist] === false
```

bcoe commented on lib/fs.js:
```diff
@@ -766,6 +774,14 @@ function mkdirSync(path, options) {
   handleErrorFromBinding(ctx);
 }
 
+Object.defineProperty(mkdirSync, features, {
```
should we also add this property on the `fs.promises` implementation?

MylesBorins replied:
I think that once we have consensus on this PR we should make a tracking issue and potentially get a bunch of different APIs properly setup with this

bcoe replied:
oh, I was more specifically thinking we should have the symbol on the `fs.promises.mkdir` function.

boneskull commented:
In the browser, conventional wisdom says "do feature checking" instead of user-agent sniffing.  For some features--such as extra, nonstandard, or unimplemented parameters--that doesn't look so pretty.  But it also means that developers tend to use up libraries (like jQuery) that do feature checking for them.

Currently, we have userland modules to handle `mkdir -p`.  The way I've seen most modules which polyfill Node.js features (e.g. `safe-buffer`) is that they'll do a feature check and [export the core module if available](https://github.com/feross/safe-buffer/blob/master/index.js#L11).

If you need e.g., `Buffer.from`, and you're not sure that your code will run in a supported environment, you use `safe-buffer`.  There's nothing unreasonable about that.

@Chalker is right that users won't want to do complex feature checking.  But they won't anyway--their libraries will do it for them.

My point is: a nontrivial feature check should not be the only reason to move the mkdirp functionality into its own function.  

A nontrivial feature check should *also* not be the only reason to create an `fs.rmdirf`; people will still use `rimraf` for a few years anyway.

bengl commented:
@bcoe Yeah for `fs.readdir(path, {withFileTypes: true}, ...)`. Right now it's discoverable on its own through the presence of a previously-not-provided constant, but that's a pretty roundabout way of getting that info.

Overall I like the currently suggested approach (`fs.mkdir[util.features].recursive`. I don't have any strong opinions on enumerability.

For `readdir` it might look something like `fs.readdir[util.features].withFileTypes`.

I think these sorts of changes should be backportable to 8.x too, which is great.

Another neat thing is that a feature-detection library (or function living in `util`!) with a friendlier interface (e.g. abstracting away the detection of the `util.features` property on a given function) could take advantage of this data.

sindresorhus commented:
I like the `fs.mkdir[util.features].recursive` approach.

I'm strongly against a `fs.mkdirp` alias, as it's too close to `fs.mkdir` and can easily be confused when reading code, and ideally there should only be one way to do something:

> There should be one-- and preferably only one --obvious way to do it. - https://www.python.org/dev/peps/pep-0020/

bengl commented on lib/fs.js:
```diff
@@ -745,6 +745,14 @@ function mkdir(path, options, callback) {
+    recursive: true
+  },
+  writable: false,
+  enumerable: false
```
If these are intended to be non-configurable, then let's make that explicit.

That being said, I think they ought to be configurable. If someone *really* wants to block this feature detection, they can always just replace `mkdir`, etc. with their own versions that don't have the property. Rather than making them go through that, might as well just make it configurable.

e.g.

```js
const realMkdir = fs.mkdir;
fs.mkdir = (...args) => realMkdir(...args);
```

MylesBorins commented:
@ChALkeR as mentioned by @boneskull I'm imagining that the majority of users would perhaps not use this. This would primarily be used by library authors, allowing mkdirp (for example) to export a wrapped version of our api rather than their implementation if it is available.

I'm personally -1 to making a mkdirp function. If we make a new function for every one of our new features that we add to modules things will explode in complexity. Building on what @bcoe said, what about `fs.rmDir`... should we have both `fs.rmDirRecursive` and `fs.rmDirRecursiveForce`?

The pattern being presented here is not solely for the case of `fs.mkdir` but will be able to be used for all other feature detection in node. While we could simply put parameters on the objects such as `fs.mkdir.recursive` this is fragile and prone to error. Having a non-enumerable non-writable known symbol will ensure that we will never accidentally write over an existing function signature in the ecosystem and that our approach cannot be monkey patched. This scales for all of node, for any of our apis

MylesBorins commented:
Added two commits

7e4ab5d moves the functionality of adding the feature detection and the creation of the symbol to `internal/util.addFeatureDetection` making adding features internally in core far less error prone. The symbol created in `internal/util` is still exposed as `util.features` but now has the description `node.util.features` as per @devsnek's request. I also added `configurable: true` to the property as per the request of @bengl .

There are also tests included in there, but we may want to expand them. If anyone wants to write more tests please feel free to push a commit on to my branch

efc9adf expands the support of the feature detection to `fs.promises` and appropriately tests / documents the change as per @bcoe's request

PTAL

MylesBorins commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16451/

jasnell commented:
I still don't think this is going to be a successful or sustainable approach but I won't block it. Just keep in mind that process.features has been in codebase for years and most core contributors don't even know about it and it's never really been used.

If I may ask, Before we add these, can we give some consideration to deprecation strategy for these? What happens if these new features need to change? What about experimental features? Will the value always be a truthy? A Boolean? Can these be nested? E.g. `thing[util.feature].something[util.feature].otherthing` or  `thing[util.feature].something.otherthing`. Before this lands, let's be sure to have even a rudimentary guideline for how they will be managed.

richardlau commented on lib/fs.js:
```diff
@@ -54,7 +54,7 @@ const {
 const { FSReqCallback, statValues } = binding;
 const internalFS = require('internal/fs/utils');
-const { getPathFromURL } = require('internal/url');
+const { addFeatureDetection, getPathFromURL } = require('internal/url');
```
`addFeatureDetection` is part of `internal/util` not `internal/url`.

MylesBorins replied:
bah... I BROKE IT

MylesBorins commented:
@mcollina the symbol is public and available at `util.features` it is also documented 😄

isaacs commented:
My $0.02

@ChALkeR makes a good point.  _Don't have any feature detection at all._

Either your app supports node versions that don't have this feature, and you have to use `mkdirp` anyway, or it doesn't, and you know it doesn't, so you just go ahead and use the builtin.

And if you wanna be fancy and support multiple versions, but also get the benefits of the faster implementation, then the feature detection belongs in the mkdirp module itself, which can look at the node version and decide what to do.

In hindsight, I agree with @jasnell that `fs.mkdirp(path)` would've probably been better than `fs.mkdir(p, {recursive: true})`.  Recursive mkdir is a different category of thing from a single mkdir call.  It's non-atomic, has a different return value, one syscall vs many, etc.  It should be a different top-level function.  (Also, terser api!)

While @MylesBorins is correct that the approach of adding a new top-level function won't scale for every API addition, it doesn't have to.  It only has to scale to the API additions which make sense to be separate functions, like this one.  (I'd argue that `fs.rimraf()` or `fs.rmrf()` would be far superior to `fs.unlink(p, { recursive: true })` or `fs.rmdir(p, { alsoRemoveChildren: true })` or some such, for the all same reasons.)

tl;dr - feature detection is a canary which has stopped singing, indicating the presence of a mistaken API.  Back up and do a different thing.

mcollina commented:
@MylesBorins I meant something created with Symbol.for(), so it works in non-node environments and places where that Symbol or module are not defined.

MylesBorins commented:
As multiple folks have mentioned `process.features` as a place to put this information I've dug up and found that it is instantiated in https://github.com/nodejs/node/blob/master/src/node.cc#L1990-L2017

It would require a bit more work to get things running as currently all features are booleans, not a deeper nested object... but that is where people would want to look if they are thinking about patching it (although we could also monkey patch process.features during instantiation (but that gives me the gibbilies).

@isaacs thanks for the 2c... I'm going to think on this for a bit.

Adding the blocked label to indicate that this requires more discussion

jasnell commented:
@MylesBorins ... to be certain, I am *not* suggesting that `process.features` is the right thing to use for this... specifically because it is mutable by userland and would suffer all of the same issues as `process.config`. Since `process.features` is not actually used by anything in core, I've been contemplating opening a PR to deprecate it.

RyanZim commented:
> then the feature detection belongs in the mkdirp module itself, which can look at the node version and decide what to do.

I agree that feature detection will mostly be in mkdirp itself, however, I do strongly disagree with mkdirp or any similar library being forced to pull in `semver` as a dependency to do this simple support check.

bcoe commented:
@RyanZim @jasnell @MylesBorins @isaacs a nice middle-ground for making features easier to detect would be having the numeric `major`, `minor` and `patch` stored somewhere on `process` -- then the feature detection logic becomes quite a bit more straight-forward:

```js
if (process.versionMajor >= 10 && process.versionMinor >= 5) {
  this.mkdirp = (path, cb) => {
     fs.mkdir(path, {recursive: true}, cb)
  }
} else {
  this.mkdirp = require('mkdirp')
}
```

tldr; I hate using regex for checking Node version #.

devsnek commented:
@bcoe i tried to land something like that twice:

https://github.com/nodejs/node/pull/19587
https://github.com/nodejs/node/pull/19438

if you can come up with something better i would love if something like that landed

RyanZim commented:
@bcoe That's certainly better if we can't have a feature detection flag.

I still dislike version sniffing because it's brittle; if `recursive` ever gets backported, libraries have to be updated to know about the backport (if there's no potential for backporting, then moot point).

rvagg commented:
ok folks, should I back mkdirp out of the v10.9.0 release going out tomorrow so we don't bake in something difficult to revert?

Trott commented:
> ok folks, should I back mkdirp out of the v10.9.0 release going out tomorrow so we don't bake in something difficult to revert?

@rvagg I think so. There's no rush on getting that feature into 10.9.0. If it waits until 10.10.0, that should be fine. That gives us the opportunity to consider changing the API as some are suggesting here. As I wrote in another issue, better to go slow and get this right then rush it out and get it wrong.

bengl commented:
Perhaps determining support for a new feature can be punted to a userland module.

I've mocked up such a module here: <https://github.com/bengl/core-features>

> **Note:** the included data is super-minimal, and just to see this working at all. Also, there's plenty of room for it being implemented in other ways.

If its data was consistently updated (and backfilled for previous features), it could be a useful alternative to building this functionality into core. In addition, this can work for older versions, prior to any version detection efforts.

Such a module *could* be adopted/blessed into the `nodejs` org, and updated as a part of the release process when any semver-minor/major commits are released. Certainly more process isn't always ideal, but I think this would be a pretty minimal addition.

mcollina commented on lib/internal/util.js:
```diff
@@ -366,8 +366,20 @@ function isInsideNodeModules() {
   return false;
 }
 
+const kCustomFeatureDetectionSymbol = Symbol('node.util.features');
```
I think this should be a `Symbol.for('node.util.features')`. It makes things a lot simpler for people maintaining isomorphic modules.

mcollina commented:
> a nice middle-ground for making features easier to detect would be having the numeric major, minor and patch stored somewhere on process -- then the feature detection logic becomes quite a bit more straight-forward

@bcoe It does just a bit, because certain features gets backported across multiple lines, which makes anything related to version checking **hard**. It works well just for semver-major changes, not for things like this.

gabrielschulhof commented:
FWIW python also takes the different-function-name approach (`os.makedirs`). @isaacs words about atomicity resonate with me. Do we want our APIs to be more like the UNIX shell utilities, or more like the system calls? I think `fs.mkdir` is a does-one-thing-and-does-it-well API. On the other hand, the name of the syscall coincides with the name of the shell utility in the UNIX world. Not so for `rm` (vs. `unlink`). I don't think that, just because the UNIX utility named `mkdir` offers recursive capabilities, ours must also.

apapirovski commented:
I agree with the above by @isaacs and @gabrielschulhof. Also the suggestion of an external package maintained and published by the org seems appealing to me (among other reasons, since it can be backwards compatible unlike something added to core).

boneskull commented:
@trott Does this block #21875 from being released?

MylesBorins commented:
I don't think we are moving forward with this, closing.

sam-github commented:
I understand why the options route to extend `fs.mkdir()` was attractive, but note that if this had been `fs.mkdirp()` the feature detection would have been `if (fs.mkdirp) ...`

Trott commented:
@sam-github This was a point raised numerous times above such as in https://github.com/nodejs/node/pull/22302#issuecomment-413036246 and https://github.com/nodejs/node/pull/22302#issuecomment-412915276.

(Ultimately, I think the thing we may be learning is that feature detection for stuff like this isn't nearly as widely needed as we imagine.)

dougwilson commented:
> Ultimately, I think the thing we may be learning is that feature detection for stuff like this isn't nearly as widely needed as we imagine.

I'm curious where the data from that came from. I mean, if there is no way to detect certain features as a module author, I'm just going to have to publish modules that are brittle and do version sniffing instead. In the past, I felt like the Node.js project said that it was bad to version sniff the `process.version`, but if that is the expected behavior to do in code, then I can definately change.

It would be awesome, though, if we're expected to version sniff, if logic such as the module `semver` was actually in core so the version sniffing is more robust than `process.version.split('.')`.

Trott commented:
> I'm curious where the data from that came from. 

It's anecdotal and I could certainly be drawing the wrong conclusion. A reasonable alternative conclusion would be "people aren't using the feature widely yet".

The data/anecdotal information is this:  We had a *lot* of debate and discussion about how to do feature detection, whether we should release the feature before we had that sorted out, etc.  When we *did* release the feature without feature detection? Nothing.

dougwilson commented:
That's just because I gave up and just assumed no feature detection would ever be delivered and just went with version sniffing. Like I said, if version sniffing is the correct solution, so be it, as it's the only option I have.
